clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetsRequestManager.java

```diff
@@ -0,0 +1,411 @@
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.BiConsumer;
+import java.util.stream.Collectors;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Manager responsible for building the following requests to retrieve partition offsets, and
+ * processing its responses.
+ * <ul>
+ *      <li>ListOffset request</li>
+ *      <li>OffsetForLeaderEpoch request</li>
+ * </ul>
+ * Requests are kept in-memory ready to be sent on the next call to {@link #poll(long)}.
+ * <br>
+ * Partition leadership information required to build ListOffset requests is retrieved from the
+ * {@link ConsumerMetadata}, so this implements {@link ClusterResourceListener} to get notified
+ * when the cluster metadata is updated.
+ */
+public class OffsetsRequestManager implements RequestManager, ClusterResourceListener {
+
+    private final ConsumerMetadata metadata;
+    private final IsolationLevel isolationLevel;
+    private final Logger log;
+    private final OffsetFetcherUtils offsetFetcherUtils;
+
+    private final Set<ListOffsetsRequestState> requestsToRetry;
+    private final List<NetworkClientDelegate.UnsentRequest> requestsToSend;
+
+    public OffsetsRequestManager(final SubscriptionState subscriptionState,
+                                 final ConsumerMetadata metadata,
+                                 final IsolationLevel isolationLevel,
+                                 final Time time,
+                                 final long retryBackoffMs,
+                                 final ApiVersions apiVersions,
+                                 final LogContext logContext) {
+        requireNonNull(subscriptionState);
+        requireNonNull(metadata);
+        requireNonNull(isolationLevel);
+        requireNonNull(time);
+        requireNonNull(apiVersions);
+        requireNonNull(logContext);
+
+        this.metadata = metadata;
+        this.metadata.addClusterUpdateListener(this);
+        this.isolationLevel = isolationLevel;
+        this.log = logContext.logger(getClass());
+        this.requestsToRetry = new HashSet<>();
+        this.requestsToSend = new ArrayList<>();
+        this.offsetFetcherUtils = new OffsetFetcherUtils(logContext, metadata, subscriptionState,
+                time, retryBackoffMs, apiVersions);
+    }
+
+    /**
+     * Determine if a there are pending fetch offsets requests to be sent and build a
+     * {@link org.apache.kafka.clients.consumer.internals.NetworkClientDelegate.PollResult}
+     * containing it.
+     */
+    @Override
+    public NetworkClientDelegate.PollResult poll(final long currentTimeMs) {
+        if (requestsToSend.isEmpty()) {
+            return new NetworkClientDelegate.PollResult(Long.MAX_VALUE, Collections.emptyList());
+        }
+
+        NetworkClientDelegate.PollResult pollResult =
+                new NetworkClientDelegate.PollResult(Long.MAX_VALUE, new ArrayList<>(requestsToSend));
+        this.requestsToSend.clear();
```

Comment:
I wonder what would happen if the request never get a response.  My main concern is retrying.  If we clear the requestsToSend collection, how do we retry if the request failed to get a response or getting sent out?

Reply:
There is a retry logic based on the `requestsToRetry`. Requests are retried if:
- the leader to get partition offsets from is not available (stale metadata case [here](https://github.com/apache/kafka/blob/34afb96b6c52617957d39ffba8fdf24f1e75bc06/clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetsRequestManager.java#L175))
- offsets were not found for a partition after trying all the known leaders, see [here](https://github.com/apache/kafka/blob/34afb96b6c52617957d39ffba8fdf24f1e75bc06/clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetsRequestManager.java#L231)

Other than that, the request will fail and won't be retried:
-  fail with error, if any, after completing requests to known leaders, [here](https://github.com/apache/kafka/blob/34afb96b6c52617957d39ffba8fdf24f1e75bc06/clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetsRequestManager.java#L235)
- timeout if no responses received (based on the ListOffsetsApplicationEvent future timeout)

This covers the retry logic of the current implementation, but let me know if you think we might be missing something.